spinnaker/spinnaker_system/google_http_lb_upsert_test.py

```diff
@@ -0,0 +1,894 @@
+    'FORWARDING_RULES': 2
+  }
+
+  MINIMUM_REGION_QUOTA = {
+      'CPUS': 2,
+      'IN_USE_ADDRESSES': 2,
+      'INSTANCE_GROUP_MANAGERS': 1,
+      'INSTANCES': 2,
+  }
+
+  @classmethod
+  def new_agent(cls, bindings):
+    '''Implements citest.service_testing.AgentTestScenario.new_agent.'''
+    agent = gate.new_agent(bindings)
+    agent.default_max_wait_secs = 600
+    return agent
+
+
+  @classmethod
+  def initArgumentParser(cls, parser, defaults=None):
+    '''Initialize command line argument parser.
+
+    Args:
+      parser: argparse.ArgumentParser
+    '''
+    super(GoogleHttpLoadBalancerTestScenario, cls).initArgumentParser(
+      parser, defaults=defaults)
+
+
+  def __init__(self, bindings, agent=None):
+    '''Constructor.
+
+    Args:
+      bindings: [dict] The data bindings to use to configure the scenario.
+      agent: [GateAgent] The agent for invoking the test operations on Gate.
+    '''
+    super(GoogleHttpLoadBalancerTestScenario, self).__init__(bindings, agent)
+
+    bindings = self.bindings
+
+    self.__lb_detail = 'lb'
+    self.TEST_APP = bindings['TEST_APP']
+    self.__lb_name = '{app}-{stack}-{detail}'.format(
+        app=bindings['TEST_APP'], stack=bindings['TEST_STACK'],
+        detail=self.__lb_detail)
+    self.__proto_hc = {
+      'name': 'basic-http-check',
+      'requestPath': '/',
+      'port': 80,
+      'checkIntervalSec': 1,
+      'timeoutSec': 1,
+      'healthyThreshold': 1,
+      'unhealthyThreshold': 1
+    }
+    self.__proto_upsert = {
+      'cloudProvider': 'gce',
+      'provider': 'gce',
+      'stack': bindings['TEST_STACK'],
+      'credentials': bindings['GCE_CREDENTIALS'],
+      'region': bindings['TEST_GCE_REGION'],
+      'loadBalancerType': 'HTTP',
+      'loadBalancerName': self.__lb_name,
+      'portRange': '80',
+      'defaultService': {
+        'name': 'default-backend-service',
+        'backends': [],
+        'healthCheck': self.__proto_hc,
+      },
+      'certificate': '',
+      'hostRules': [
+        {
+          'hostPatterns': ['host1.com', 'host2.com'],
+          'pathMatcher': {
+            'pathRules': [
+              {
+                'paths': ['/path', '/path2/more'],
+                'backendService': {
+                  'name': 'backend-service',
+                  'backends': [],
+                  'healthCheck': self.__proto_hc,
+                }
+              }
+            ],
+            'defaultService': {
+              'name': 'pm-backend-service',
+              'backends': [],
+              'healthCheck': self.__proto_hc,
+            }
+          }
+        }
+      ],
+      'type': 'upsertLoadBalancer',
+      'availabilityZones': {bindings['TEST_GCE_REGION']: []},
+      'user': '[anonymous]'
+    }
+
+
+  def _make_bs_link(self, bs):
+    '''Make a fully-formatted backend service link.
+    '''
+    return (GCE_URL_PREFIX + self.bindings['GCE_PROJECT'] +
+            '/global/backendServices/' + bs)
+
```

Comment:
I prefer "get_*_link". Your other "make" have side effects or are more factory like. This is conceptually a lookup or derivation.